[JSC] Reduce more samples in DFG finalization

Constellation opened the pull request:
#### 5b30b2ee35e0f4315ce4f9b41fcc5277097498dd
<pre>
[JSC] Reduce more samples in DFG finalization
<a href="https://bugs.webkit.org/show_bug.cgi?id=272925">https://bugs.webkit.org/show_bug.cgi?id=272925</a>
<a href="https://rdar.apple.com/126711361">rdar://126711361</a>

Reviewed by Keith Miller.

This patch attempts to remove more samples in DFG finalization, observed in performance profiling.

1. Pass mustHandleValues via WTFMove to avoid additional allocation.
2. Clean up completeAllReadyPlansForVM to avoid taking a lock twice.

* Source/JavaScriptCore/dfg/DFGDriver.cpp:
(JSC::DFG::compileImpl):
(JSC::DFG::compile):
* Source/JavaScriptCore/dfg/DFGDriver.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::tierUpCommon):
* Source/JavaScriptCore/dfg/DFGPlan.cpp:
(JSC::DFG::Plan::Plan):
* Source/JavaScriptCore/dfg/DFGPlan.h:
* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/jit/JITWorklist.cpp:
(JSC::JITWorklist::completeAllReadyPlansForVM):
(JSC::JITWorklist::cancelAllPlansForVM):
(JSC::JITWorklist::removeAllReadyPlansForVM):
* Source/JavaScriptCore/jit/JITWorklist.h:

Canonical link: <a href="https://commits.webkit.org/277753@main">https://commits.webkit.org/277753@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0de692c5e7863783f98f9ef78c8dbb609bb56330

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48474 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27686 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/51434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51162 "Failed to checkout and rebase branch from PR 27471") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/44539 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33622 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25209 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/51162 "Failed to checkout and rebase branch from PR 27471") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49056 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/25367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/51434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/51162 "Failed to checkout and rebase branch from PR 27471") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/22856 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/51434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6531 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/41771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/44792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/51434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53067 "Built successfully") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/47965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/23521 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/19845 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/46952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/24786 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/51434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/45864 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/25591 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/55460 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6901 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/24509 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/11406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/jit/JITWorklist.cpp:
```diff
@@ -319,21 +302,32 @@ void JITWorklist::dump(const AbstractLocker& locker, PrintStream& out) const
+        return true;
+    });
+
+    if (!!requestedKey) {
```
Nit: We could just have an `explicit operator bool()` on `JITCompilationKey` instead of `bool operator!` so you wouldn't need `!!` here.

Constellation replied:
Nice, changed.

kmiller68 reviewed (approved):
r=me.

Constellation committed:
[JSC] Reduce more samples in DFG finalization
https://bugs.webkit.org/show_bug.cgi?id=272925
rdar://126711361

Reviewed by Keith Miller.

This patch attempts to remove more samples in DFG finalization, observed in performance profiling.

1. Pass mustHandleValues via WTFMove to avoid additional allocation.
2. Clean up completeAllReadyPlansForVM to avoid taking a lock twice.

* Source/JavaScriptCore/dfg/DFGDriver.cpp:
(JSC::DFG::compileImpl):
(JSC::DFG::compile):
* Source/JavaScriptCore/dfg/DFGDriver.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::tierUpCommon):
* Source/JavaScriptCore/dfg/DFGPlan.cpp:
(JSC::DFG::Plan::Plan):
* Source/JavaScriptCore/dfg/DFGPlan.h:
* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/jit/JITWorklist.cpp:
(JSC::JITWorklist::completeAllReadyPlansForVM):
(JSC::JITWorklist::cancelAllPlansForVM):
(JSC::JITWorklist::removeAllReadyPlansForVM):
* Source/JavaScriptCore/jit/JITWorklist.h:

Canonical link: https://commits.webkit.org/277753@main
